### PR TITLE
Introduce ErrorProneRuntimeClasspath check

### PR DESCRIPTION
**Pull request by Stephan202**
~:exclamation: This PR is on top of #881. :exclamation:~

Suggested commit message:
```
Introduce `ErrorProneRuntimeClasspath` check (#882)

Prefer "type-safe" type references were possible, but use string
literals if the references type may not be available at runtime.
```

The PR is split into three commits that can best be reviewed one-by-one: the check, its automated suggestions, and manual tweaks.

### CODE REVIEW

**Comment by Stephan202** [comment]
I've started to resolve the conflicts here. Almost done; will finalize after some meetings.

**Comment by Stephan202** [comment]
Rebased; the PR now targets `master` and is ready for review.

**Comment by Stephan202** [comment]
Once again rebased and resolved a conflict. @rickie and @mohamedsamehsalah I suspect this'll start happening more ;)

**Review by mohamedsamehsalah — APPROVED** [review]
Nothing stands out for me :octocat: 

Is there a general rule to follow when determining if a class is available on run-time or not ?

**Review by Stephan202 — COMMENTED** [review]
I added one more commit, in which I resorted a few things. The only place I didn't reorder is the listing in `JUnitValueSource`, as there the existing order seems more intuitive.

> Is there a general rule to follow when determining if a class is available on run-time or not ?

Thanks for reviewing @mohamedsamehsalah! I'm not sure about a "general rule", but we use [this code](https://github.com/PicnicSupermarket/error-prone-support/blob/cc36aa993c7c7aedb5fc5c18dc829bd7a50ca9b3/error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/util/ThirdPartyLibrary.java#L73-L100) to check whether a given class is "currently" on the classpath. That of course doesn't mean that a class will also be on the classpath when the code is executed at a later point in time. For example, we have a bunch of `provided` dependencies that may not _actually_ be "provided" by downstream users. That's why the `ErrorProneRuntimeClasspath` check restricts itself to types that match the `CLASSPATH_TYPES` regex: those are known to be dependencies of Error Prone, and thus ~guaranteed to be present. I've also added a comment making this more explicit :)

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ErrorProneRuntimeClasspath.java [hunk 1]
```diff
@@ -59,16 +59,22 @@
...
-      Pattern.compile(
-          "^(com\\.google\\.common\\..*|com\\.google\\.errorprone\\.([^\\.]+(?<!TestHelper)(\\..*)?)|java\\..*)$");
   private static final Matcher<ExpressionTree> GET_CANONICAL_NAME_INVOCATION =
       instanceMethod().onExactClass(Class.class.getCanonicalName()).named("getCanonicalName");
   private static final ImmutableSet<String> PRIMITIVE_TYPES =
       Primitives.allPrimitiveTypes().stream()
           .map(Class::getCanonicalName)
           .collect(toImmutableSet());
 
+  /**
```
Comment (mohamedsamehsalah) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ErrorProneRuntimeClasspath.java:69]: ❤️

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ErrorProneRuntimeClasspath.java [hunk 2]
```diff
@@ -59,16 +59,22 @@
...
+   * A pattern that matches fully qualified type names that are expected to be runtime dependencies
```
Comment (mohamedsamehsalah) [inline error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ErrorProneRuntimeClasspath.java:70]: ❤️

**Review by rickie — COMMENTED** [review]
Cool PR, sorry for delay in review :). 🚀

**Review by rickie — APPROVED** [review]
Added the commit as it can probably be improved. Other issue is not a blocker for me so already approving.

File: error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ErrorProneRuntimeClasspath.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+    return fix.replace(
+            original,
+            identifier
+                + ".class.getCanonicalName()"
+                + (suffix.isEmpty() ? "" : (" + " + Constants.format(suffix))))
+        .build();
+  }
+
+  @Override
+  public Description matchMethodInvocation(MethodInvocationTree tree, VisitorState state) {
```
[the remark this answers is not in the document]
Comment (Stephan202) [reply error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/ErrorProneRuntimeClasspath.java:103]: Indeed. Here I think the idea was to have two "groups" methods, one for each "direction", but I see now that this is not how it turned out. Will move :+1:

**Review by Stephan202 — COMMENTED** [review]
Rebased and added a commit.

Ideally we move this check, and some like it, to a separate Maven module, so that downstream users won't have to disable it. But that's something we can/should pick up separately.

**Pull request merged**